### PR TITLE
Image upload & display

### PR DESCRIPTION
**Pull request by AniMachaidze**
Added image upload using Blobstore API and display functionality in comments.

image-comment

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -33,10 +33,22 @@
...
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.util.Map;
+import java.net.MalformedURLException;
+import java.net.URL;
```
Comment (oydang) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:46]: Nit: alphabetize your import statements

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -151,6 +165,42 @@ private String getParameter(HttpServletRequest request, String name,
...
+        }
+
+        ImagesService imagesService = ImagesServiceFactory.getImagesService();
+        ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+        try {
+            URL url = new URL(imagesService.getServingUrl(options));
+            return url.getPath();
+        } catch (MalformedURLException e) {
+            return imagesService.getServingUrl(options);
```
Comment (oydang) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:200]: and Nit: Add a log statement here explaining the exception and how it's handled

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -158,4 +162,16 @@ function login() {
...
+}
+
+function blobUpload() {
+	fetch('/blobstore-upload-url')
+		.then((response) => {
+			return response.text();
+		})
+		.then((imageUploadUrl) => {
+			const commentForm = document.getElementById('form');
+			commentForm.action = imageUploadUrl;
```
Comment (oydang) [inline portfolio/src/main/webapp/script.js:175]: So, does this successfully change the form action on career.html to upload an image? After this happens, can the user still use the form to add a comment? I don't really understand how setting the action to an upload URL actually works.

Comment (AniMachaidze) [reply portfolio/src/main/webapp/script.js:175]: BlobStoreServlet is creating uploadUrl which has succespath "/data", uploadUrl becomes new action for the form. The way I understand, when submit button is clicked, the form first redirects to uploadUrl which is used by Blobstore API, it processes file and adds info about it to request object. After successfully doing that, it redirects to "/data" with the request that has info about blob. From here, as before, we can get each parameter of the form and get uploaded image URL from request object  (which is handled by method getUploadedFileUrl)

Comment (oydang) [reply portfolio/src/main/webapp/script.js:175]: Got it, thank you for explaining that!

File: portfolio/pom.xml [hunk 4]
```diff
@@ -46,6 +46,10 @@
         <artifactId>appengine-maven-plugin</artifactId>
         <version>2.2.0</version>
         <configuration>
+<<<<<<< HEAD
+          <!-- TODO: set project ID.-->
+=======
+>>>>>>> ddfd09d5ebdb7eebc0e86472254069ae99a82a17
```
Comment (oydang) [inline portfolio/pom.xml:52]: You'll need to resolve this merge conflict.

**Pull request merged**